File: 03. hash_table/01. two_sum.py
"""
nums = [3,2,4],
target = 6

map = {
    3 : 0
    2 : 1
}

required = 6 - 3 => 3
3 in map ? no

required = 6 - 2 => 4
4 in map ? no

required = 6 - 4 => 2
2 in map ? Yes


"""

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        map = {}

        for i in range(n):
            required = target - nums[i]
            if required in map:
                return [i, map[required]]
            map[nums[i]] = i

File: 06. hash_table/01. Two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  const map = {};
  for (let i = 0; i < nums.length; i++) {
    const currentMapVal = map[nums[i]];
    if (currentMapVal >= 0) {
      return [currentMapVal, i];
    } else {
      const required = target - nums[i];
      map[required] = i;
    }
  }
  return null;
};

const nums = [2, 7, 11, 15];
const target = 9;
console.log(twoSum(nums, target));
